### ISSUE TITLE
LanguageClientImpl is required but is not available

### ISSUE THREAD

**Issue by vgcpge** [eclipse-lsp4e/lsp4e#626]
Extension point `org.eclipse.lsp4e.languageServer` requires a class given in the field `clientImpl` to extends `org.eclipse.lsp4e.LanguageClientImpl`. However `org.eclipse.lsp4e` package is internal and an attempt to extend the class leads to warnings.

Thus, extensions for lsp4e are unable to customize the client behavior or schema in a legitimate manner.

There are following solutions available:
1. Require `clientImpl` to implement `org.eclipse.lsp4j.services.LanguageClient` interface instead.
1.1 (Optional) Provide another basic implementation of  `org.eclipse.lsp4j.services.LanguageClient` for extensions to reuse.
2. Make `org.eclipse.lsp4e` an API package (too many classes in that package to expose them as API).
3. Move `org.eclipse.lsp4e.LanguageClientImpl` to another package (backward incompatible change).

Context: I've been trying to implement [partial result progress with data transfer](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#partialResults) in an extension and default client does  partial progress reporting, but does not expose the data collected.

**Comment by sebthom** [comment]
@rubenporras @mickaelistria any thoughts on this. I am faced with the same issue and so far simply ignore the fact that it is an internal/restricted class.

**Comment by mickaelistria** [comment]
The overall proposal seems fine; here his some nitpicking bits of mine (which of course are not necessarily best; as it's about impossible to define what "best" mean when designing APIs; as one cannot really know all the possible use-cases already in place)

* HasLanguageServerPropertyTester is not expected to be referenced directly from Java code but is supposed to be used with the property tester in plugin.xml.
* LanguageServiceAccessor is kind of internal, as `LanguageServers` is now a better entry point to get the right LS
* `LaunchConfigurationStreamProvider` should IMO be internal as it's not really meant to be extensible; it's more an internal shortcut for quickly testing a LS.
* LoggingStreamConnectionProviderProxy should also be internal as it's about the internal Logging capabilities of LSP4E that are not supposed to be API
* LanguageServersRegistry and ContentTypeTo* classes are not meant to be used by extenders AFAIK

**Comment by sebthom** [comment]
To be honest, I'm not really comfortable with making `LanguageServers`, `LanguageServerWrapper`, `LanguageServiceAccessor`, and `LanguageServersRegistry` officially accessible like this at this point. The current API feels quite fragmented, and I think it would benefit from refactoring/consolidation before being exposed more broadly. Ideally, we wouldn’t need four separate classes as entry points to discover and work with a language server.

**Comment by mickaelistria** [comment]
I tend to agree with @sebthom : instead of just moving things here and there; maybe the project would better benefit from an effort of redesigning APIs, gathering input from main consumers to build minimal yet powerful APIs that match the expectations.
Such API design has never been fully done: when LSP4E incepted, the goal was to make it work, then it grew into being higher functional quality with good logging/debug assistance, then it grew into becoming a framework more than just a client... Nowadays, the scope seems more stable and there are enough consumers to design a proper API. So it's a good time to reconsider the API from scratch.

**Comment by basilevs** [comment]
When moving, it would be nice to remove Impl suffix to indicate the class is an API and is extensible.

**Comment by sebthom** [comment]
> Maybe I have just had bad experiences, but I think major redesigns are very hard to get right, and prefer more the evolutionary approach. So, for example, since I think we all agree that `LanguageClientImpl` should be public for extension, what about we create a new package, for example `org.eclipse.lsp4e.client`, and move it there?
> 
I am not concerned about LanguageClientImpl. I think this can be public as-is. 

Only the LanguageServerWrapper/LanguageServiceAccessor/LanguageServers trio gives me a bit of a headache.

**Comment by rubenporras** [comment]
> [@rubenporras](https://github.com/rubenporras) I am not concerned about LanguageClientImpl. I think this can be public as-is.

I also think so. My reasoning is that since 

- we are not concerned at all about `LanguageClientImpl` but we are concerned about other classes in `org.eclipse.lsp4e`. 
- personally I am very afraid of a big API refactor

the easy way out is to rename and move `LanguageClientImpl` to a package we can make public, then refactor `LanguageClientImpl` to use the new class and mark it as deprecated.

This is something any of us can easily do. I also believe that with this small extractions from `org.eclipse.lsp4e` we can get very far until a big API refactor is needed, and I would be much more confident then.

**Comment by sebthom** [comment]
we could move the LanguageClientImpl to a new public api or spi package and keep a deprecated LanguageClientImpl that extends the new public one for e. g. one or two releases and then remove it.

**Comment by sebthom** [comment]
What if we would do it the otherway around. I.e. Leave LanguageClientImpl where it is, make that package public and mark all classes we don't want to be public as deprecated for removal?

**Comment by rubenporras** [comment]
I think that creates a lot of noise for any current consumer. Actually, after looking at the change to extract the class to a new package, I think it is quite fine.

I have created one PR and set @sebthom and @mickaelistria. I have just not renamed the class, since after thinking about it, the current name looks okay to me.
